### PR TITLE
Fix bug in TTL merges assignment

### PR DESCRIPTION
**Pull request by alesapin**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Bug Fix

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix bug which can lead to intersecting parts after merges with TTL: `Part all_40_40_0 is covered by all_40_40_1 but should be merged into all_40_41_1. This shouldn't happen often.`

### CODE REVIEW

**Comment by alesapin** [comment]
Wow! Flaky check possibly reproduced the bug. I'll try with thread fuzzer.

**Comment by alesapin** [comment]
```
2021.06.21 21:04:28.827229 [ 1455 ] {} <Error> test_vofhux.ttl_table2 (1de274a3-4e40-4ffd-9de2-74a34e401ffd): auto DB::StorageReplicatedMergeTree::processQueueEntry(ReplicatedMergeTreeQueue::SelectedEntryPtr)::(anonymous class)::operator()(DB::StorageReplicatedMergeTree::LogEntryPtr &) const: Code: 49, e.displayText() = DB::Exception: Part all_24_24_0 is covered by all_24_24_1 but should be merged into all_24_25_1. This shouldn't happen often., Stack trace (when copying this message, always include the lines below):

0. ./obj-x86_64-linux-gnu/../contrib/libcxx/include/exception:133: Poco::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) @ 0x2a4bf149 in /usr/bin/clickhouse
1. ./obj-x86_64-linux-gnu/../src/Common/Exception.cpp:59: DB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int, bool) @ 0xb446042 in /usr/bin/clickhouse
2. ./obj-x86_64-linux-gnu/../contrib/libcxx/include/string:1444: DB::Exception::Exception<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&>(int, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) @ 0x220fda4d in /usr/bin/clickhouse
3. ./obj-x86_64-linux-gnu/../src/Storages/StorageReplicatedMergeTree.cpp:1646: DB::StorageReplicatedMergeTree::tryExecuteMerge(DB::ReplicatedMergeTreeLogEntry const&) @ 0x21f7fb87 in /usr/bin/clickhouse
4. ./obj-x86_64-linux-gnu/../src/Storages/StorageReplicatedMergeTree.cpp:0: DB::StorageReplicatedMergeTree::executeLogEntry(DB::ReplicatedMergeTreeLogEntry&) @ 0x21f5b979 in /usr/bin/clickhouse
5. ./obj-x86_64-linux-gnu/../contrib/libcxx/include/functional:0: bool std::__1::__function::__policy_invoker<bool (std::__1::shared_ptr<DB::ReplicatedMergeTreeLogEntry>&)>::__call_impl<std::__1::__function::__default_alloc_func<DB::StorageReplicatedMergeTree::processQueueEntry(std::__1::shared_ptr<DB::ReplicatedMergeTreeQueue::SelectedEntry>)::$_18, bool (std::__1::shared_ptr<DB::ReplicatedMergeTreeLogEntry>&)> >(std::__1::__function::__policy_storage const*, std::__1::shared_ptr<DB::ReplicatedMergeTreeLogEntry>&) @ 0x220f40d2 in /usr/bin/clickhouse
6. ./obj-x86_64-linux-gnu/../src/Storages/MergeTree/ReplicatedMergeTreeQueue.cpp:1351: DB::ReplicatedMergeTreeQueue::processEntry(std::__1::function<std::__1::shared_ptr<zkutil::ZooKeeper> ()>, std::__1::shared_ptr<DB::ReplicatedMergeTreeLogEntry>&, std::__1::function<bool (std::__1::shared_ptr<DB::ReplicatedMergeTreeLogEntry>&)>) @ 0x22a42ae7 in /usr/bin/clickhouse
7. ./obj-x86_64-linux-gnu/../src/Storages/StorageReplicatedMergeTree.cpp:0: DB::StorageReplicatedMergeTree::processQueueEntry(std::__1::shared_ptr<DB::ReplicatedMergeTreeQueue::SelectedEntry>) @ 0x22007d12 in /usr/bin/clickhouse
8. ./obj-x86_64-linux-gnu/../src/Storages/MergeTree/BackgroundJobsExecutor.cpp:103: void std::__1::__function::__policy_invoker<void ()>::__call_impl<std::__1::__function::__default_alloc_func<DB::IBackgroundJobExecutor::jobExecutingTask()::$_0, void ()> >(std::__1::__function::__policy_storage const*) @ 0x22465d04 in /usr/bin/clickhouse
9. ./obj-x86_64-linux-gnu/../contrib/libcxx/include/functional:0: ThreadPoolImpl<ThreadFromGlobalPool>::worker(std::__1::__list_iterator<ThreadFromGlobalPool, void*>) @ 0xb4e2b4c in /usr/bin/clickhouse
10. ./obj-x86_64-linux-gnu/../src/Common/ThreadPool.cpp:0: ThreadFromGlobalPool::ThreadFromGlobalPool<void ThreadPoolImpl<ThreadFromGlobalPool>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda0'()>(void&&, void ThreadPoolImpl<ThreadFromGlobalPool>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda0'()&&...)::'lambda'()::operator()() @ 0xb4e8239 in /usr/bin/clickhouse
11. ./obj-x86_64-linux-gnu/../contrib/libcxx/include/functional:0: ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) @ 0xb4dd40c in /usr/bin/clickhouse
12. ./obj-x86_64-linux-gnu/../contrib/libcxx/include/memory:1655: void* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda0'()> >(void*) @ 0xb4e4bb4 in /usr/bin/clickhouse
13. start_thread @ 0x9609 in /usr/lib/x86_64-linux-gnu/libpthread-2.31.so
14. clone @ 0x122293 in /usr/lib/x86_64-linux-gnu/libc-2.31.so
 (version 21.7.1.7236)
```

**Comment by alesapin** [comment]
@Mergifyio update

File: src/Storages/MergeTree/MergeTreeDataMergerMutator.cpp [hunk 1]
```diff
@@ -264,19 +264,32 @@ SelectPartsDecision MergeTreeDataMergerMutator::selectPartsToMerge(
...
                 assert(!parts_ranges.back().empty());
-                parts_ranges.emplace_back();
 
-                /// Now we have no previous part, but it affects only logging
-                prev_part = nullptr;
+                /// Some parts cannot be merged with previous parts and also cannot be merged with themselves,
+                /// for example, merge is already assigned for such parts, or they participate in quorum inserts
+                /// and so on.
+                /// Also we don't start new interval here (maybe all next parts cannot be merged and we don't want to have empty interval)
+                if (!can_merge_callback(nullptr, part, nullptr))
```
Comment (tavplubix) [inline src/Storages/MergeTree/MergeTreeDataMergerMutator.cpp:288]: Do we need to start new interval in this case?

Comment (tavplubix) [reply src/Storages/MergeTree/MergeTreeDataMergerMutator.cpp:288]: Ok, seems like it's done here
https://github.com/ClickHouse/ClickHouse/blob/c15493f0df3dc577e22760553eff452c07a83fba/src/Storages/MergeTree/MergeTreeDataMergerMutator.cpp#L269-L270

**Review by tavplubix — APPROVED** [review]

**Comment by alesapin** [comment]
Is it my fix or the test show this error?
```
2021.06.22 20:24:21.778684 [ 647 ] {} <Error> test_kd1w0m.ttl_table4 (4d173b4f-7696-4528-8d17-3b4f7696e528): auto DB::StorageReplicatedMergeTree::processQueueEntry(ReplicatedMergeTreeQueue::SelectedEntryPtr)::(anonymous class)::operator()(
DB::StorageReplicatedMergeTree::LogEntryPtr &) const: Code: 49, e.displayText() = DB::Exception: Unexpected merged part all_206_214_1 intersecting drop range all_208_209_999999999, Stack trace (when copying this message, always include the
 lines below):                                                                                                                                                                                                                                 
                                                                                                                                                                                                                                               
0. ./obj-x86_64-linux-gnu/../contrib/libcxx/include/exception:133: Poco::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) @ 0x2a4c8889 in /usr/bin/clickhouse           
1. ./obj-x86_64-linux-gnu/../src/Common/Exception.cpp:59: DB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int, bool) @ 0xb44f182 in /usr/bin/clickhouse                 
2. DB::Exception::Exception<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >(int, std::__1::basic_string<c
har, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocat
or<char> >&&) @ 0x1b0903a5 in /usr/bin/clickhouse
3. ./obj-x86_64-linux-gnu/../src/Storages/MergeTree/MergeTreeData.cpp:2339: DB::MergeTreeData::removePartsInRangeFromWorkingSet(DB::MergeTreePartInfo const&, bool, std::__1::unique_lock<std::__1::mutex>&) @ 0x225c53bd in /usr/bin/clickhous
e
4. ./obj-x86_64-linux-gnu/../src/Storages/StorageReplicatedMergeTree.cpp:0: DB::StorageReplicatedMergeTree::executeDropRange(DB::ReplicatedMergeTreeLogEntry const&) @ 0x21f68794 in /usr/bin/clickhouse
5. ./obj-x86_64-linux-gnu/../src/Storages/StorageReplicatedMergeTree.cpp:0: DB::StorageReplicatedMergeTree::executeLogEntry(DB::ReplicatedMergeTreeLogEntry&) @ 0x21f62386 in /usr/bin/clickhouse
6. ./obj-x86_64-linux-gnu/../contrib/libcxx/include/functional:0: bool std::__1::__function::__policy_invoker<bool (std::__1::shared_ptr<DB::ReplicatedMergeTreeLogEntry>&)>::__call_impl<std::__1::__function::__default_alloc_func<DB::Storag
eReplicatedMergeTree::processQueueEntry(std::__1::shared_ptr<DB::ReplicatedMergeTreeQueue::SelectedEntry>)::$_18, bool (std::__1::shared_ptr<DB::ReplicatedMergeTreeLogEntry>&)> >(std::__1::__function::__policy_storage const*, std::__1::sha
red_ptr<DB::ReplicatedMergeTreeLogEntry>&) @ 0x220fd6d2 in /usr/bin/clickhouse
7. ./obj-x86_64-linux-gnu/../src/Storages/MergeTree/ReplicatedMergeTreeQueue.cpp:1351: DB::ReplicatedMergeTreeQueue::processEntry(std::__1::function<std::__1::shared_ptr<zkutil::ZooKeeper> ()>, std::__1::shared_ptr<DB::ReplicatedMergeTreeL
ogEntry>&, std::__1::function<bool (std::__1::shared_ptr<DB::ReplicatedMergeTreeLogEntry>&)>) @ 0x22a4c727 in /usr/bin/clickhouse
8. ./obj-x86_64-linux-gnu/../src/Storages/StorageReplicatedMergeTree.cpp:0: DB::StorageReplicatedMergeTree::processQueueEntry(std::__1::shared_ptr<DB::ReplicatedMergeTreeQueue::SelectedEntry>) @ 0x22011312 in /usr/bin/clickhouse
9. ./obj-x86_64-linux-gnu/../src/Storages/MergeTree/BackgroundJobsExecutor.cpp:103: void std::__1::__function::__policy_invoker<void ()>::__call_impl<std::__1::__function::__default_alloc_func<DB::IBackgroundJobExecutor::jobExecutingTask()
::$_0, void ()> >(std::__1::__function::__policy_storage const*) @ 0x2246f304 in /usr/bin/clickhouse
10. ./obj-x86_64-linux-gnu/../contrib/libcxx/include/functional:0: ThreadPoolImpl<ThreadFromGlobalPool>::worker(std::__1::__list_iterator<ThreadFromGlobalPool, void*>) @ 0xb4ebc8c in /usr/bin/clickhouse
11. ./obj-x86_64-linux-gnu/../src/Common/ThreadPool.cpp:0: ThreadFromGlobalPool::ThreadFromGlobalPool<void ThreadPoolImpl<ThreadFromGlobalPool>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lamb
da0'()>(void&&, void ThreadPoolImpl<ThreadFromGlobalPool>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda0'()&&...)::'lambda'()::operator()() @ 0xb4f1379 in /usr/bin/clickhouse
12. ./obj-x86_64-linux-gnu/../contrib/libcxx/include/functional:0: ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) @ 0xb4e654c in /usr/bin/clickhouse
13. ./obj-x86_64-linux-gnu/../contrib/libcxx/include/memory:1655: void* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, void ThreadPoolImpl<std
::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda0'()> >(void*) @ 0xb4edcf4 in /usr/bin/clickhouse
14. start_thread @ 0x9609 in /usr/lib/x86_64-linux-gnu/libpthread-2.31.so
15. __clone @ 0x122293 in /usr/lib/x86_64-linux-gnu/libc-2.31.so
 (version 21.7.1.7248)
```

**Comment by tavplubix** [comment]
But there is no `ALTER PARTITION` queries in the test...

**Comment by tavplubix** [comment]
It can be related to background removal of empty parts

**Comment by tavplubix** [comment]
Yes, the following code is incorrect:
https://github.com/ClickHouse/ClickHouse/blob/67e9b85951d47ba2c056f3cb8daaf1cf3459d869/src/Storages/StorageReplicatedMergeTree.cpp#L6865-L6872
https://github.com/ClickHouse/ClickHouse/blob/67e9b85951d47ba2c056f3cb8daaf1cf3459d869/src/Storages/StorageReplicatedMergeTree.cpp#L3163-L3166

It must check for covering virtual part

**Comment by alesapin** [comment]
No, my fix incorrect again...

**Comment by alesapin** [comment]
Nice, timeouts are expected for the flaky check. I'll try to make this test slightly more lightweight.

**Comment by alesapin** [comment]
This test just too heavy for a concurrent run in flaky check... I'll try to remove sync replica.

**Comment by alesapin** [comment]
Still too heavy. I'll remove it from the parallel list.

**Comment by alesapin** [comment]
Failures not related to changes.

**Pull request merged**